### ISSUE TITLE
Tutorial DataModule leads to Validation data leakage in Multiprocessing

### ISSUE THREAD

**Issue by profPlum** [Lightning-AI/lightning#18638]
### 📚 Documentation

It is recommended in the docs to do splits in the setup() method & [an example is given](https://pytorch-lightning.readthedocs.io/en/1.7.7/data/datamodule.html?highlight=random_split#what-is-a-datamodule) that does random splits in setup.

Perhaps it shouldn't be recommended to do random splits in the setup(). Because 
I've found that when I do random_split inside setup with multiprocessing it results in each GPU worker having a different validation_dataset (given by random split).

<img width="826" alt="image" src="https://github.com/Lightning-AI/lightning/assets/4739054/f457ca72-e9f8-442e-96e7-1bb1c0b78ff2">

_So when training the model is trained on some (or all) of the validation data when random splits are done in setup()._ (without a seed)

cc @borda @carmocca @awaelchli

**Comment by awaelchli** [comment]
IMO if users think this is not clear, the example should be updated to set the generator on the [`random_split` function with a fixed seed](https://pytorch.org/docs/stable/data.html#torch.utils.data.random_split). 

However, I would like to defend the example in that it is a code snippet that tries to highlight the following point:

> The equivalent DataModule just organizes the same exact code, but makes it reusable across projects.

It doesn't try to give you a full end-to-end training example but rather just get this message across so you can build your own.

**Comment by profPlum** [comment]
@awaelchli I don't think its a good idea to encourage & even demonstrate code smells  which can easily lead to serious (& silent) bugs later on. What if someone started their projected based off a docs example then scaled to multiple GPUs & went to publish their results without reading the finer points of setup() vs prepare_data()?

Moreover maybe you shouldn't be conveying that you can just copy and paste when the reality is quite different for multi-gpu training? Also there are many other places in the docs which encourage splits in setup(). I think all of these places should _heavily_ emphasize this GOTCHA. Honestly I'm not even sure what the best practice to handle this seemingly common issue would be? Docs should mention best practice with the GOTCHA.

Or even better perhaps an elegant design change could solve the problem? It seems like the point in lightning is so researchers don't have to worry about this sort of thing. But honestly I'm not 100% sure what that would look like? Maybe safe abstraction of validation/test splits (e.g. keras style), idk.

**Comment by awaelchli** [comment]
Has my last reply left the wrong impression @profPlum? I'm sorry if it has frustrated you but I have never said no to any of your comments and I have not objected to anything you said. I simply stated my opinion in addition to this ticket. 

You are proposing that we don't recommend doing the split in setup(), but you haven't said where it should be done otherwise. Could you clarify please? Would my suggestion of setting the random seed on the random_split function not work and why not?

**Comment by profPlum** [comment]
I'm sorry too, perhaps I conveyed the wrong tone. And you're right I'm not 100% sure of perfect solution either...

I'm just worried that your response was too narrow in regards to changing one line of code. Whereas the problem is more widespread; it would be better to change every similar part of the documentation & clearly emphasize the GOTCHA + best practice.

Honestly it still scares me that _even after docs_ change it will be so easy to make such a serious mistake which could easily go undetected.
Ideally: Things would be designed without GOTCHAs, & so that users would simply be unable to make this kind of mistake. But I'm not sure how to do that.

**Comment by profPlum** [comment]
@awaelchli But in regards to doing it in setup vs somewhere else I was thinking it would be nice if it were easy to do shuffling/splitting in prepare_data(). But I'm not sure how to do that in a way which is safe when training multiple models at once in a shared file system (e.g. slurm/cloud). Without directly coupling the code with slurm or something else.

**Comment by awaelchli** [comment]
Sure, we can make that clearer in the docs. 

I'm struggling to understand though how and why Lightning should enforce such behavior. It doesn't fit well with the core philosophy. 

`prepare_data()` is definitely not the place to create the split, not when it's done online. However, if you do the split once and then save to disk in prepare_data, that's ok and you can then load it in setup(). You see, this is very dependent on the user's dataset and use case and so we need to leave the flexibility for the user to make this choice which hook is more suited. 

I really don't think Lightning should change the default behavior of torch functions that the user calls, nor should it silently set the seed or random state if the user doesn't explicitly ask for it. Lightning's philosophy is to organize PyTorch code and provide good default configuration, but it does not in any way intend to override the behavior of torch on a functional level.

**Comment by profPlum** [comment]
@awaelchli Like I said I don't know of a perfect solution & I'm not recommending you change behavior of pytorch functions.

Out of curiosity how is this problem dealt with in raw pytorch in a multiprocessing setting? There must be some kind of best practice associated with random_split.

It would be surprising if that practice were to make the split deterministic since it seems to defeat the purpose of the function.

**Comment by profPlum** [comment]
Oh interesting, ya I was getting that impression since keras also does deterministic splitting. I guess then encouraging seeding is probably our best bet, for now.

**Comment by profPlum** [comment]
@awaelchli Hi again, it just occurred to me that this problem goes away with CLI arg `--seed_everything=$RANDOM`. But also I checked the default value for this arg & it's `--seed_everything=True`. I honestly can't tell why these two settings would behave differently: `--seed_everything=$RANDOM` & `--seed_everything=True` (default).

My understanding then would be that by default the problem I mentioned shouldn't be a problem (given the current defaults). But it is, why is that?

**Comment by awaelchli** [comment]
@profPlum In addition, could you explain how you are launching the script? ``--seed_everything=True``, if we're talking about the LightningCLI, will make it so that a random number is chosen for the seed. The best practice is however to always pass in a fixed number, so that each process has the same seed from the beginning.

**Comment by awaelchli** [comment]
@profPlum Let me know what you think about my last reply and the PR I sent.

**Comment by profPlum** [comment]
> @profPlum Let me know what you think about my last reply and the PR I sent.

Looks great!
